# Project Description

The objective of this project is to help finding the best place for a new well for XYZ mining company.

Steps to choose the location:
- Collect the oil well parameters in the selected region: oil quality and volume of reserves;
- Build a model for predicting the volume of reserves in the new wells;
- Pick the oil wells with the highest estimated values;
- Pick the region with the highest total profit for the selected oil wells.
We have data on oil samples from three regions. Parameters of each oil well in the region are already known. We are build a model that will help to pick the region with the highest profit margin.

## Importing necessary libraries

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score 
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from scipy import stats as st


## Importing and Preparing the data

**Importing and studying Region 1 data**

In [2]:
data1 = pd.read_csv('/datasets/geo_data_0.csv')
data1.head(10)

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
5  wX4Hy  0.969570  0.489775 -0.735383   64.741541
6  tL6pL  0.645075  0.530656  1.780266   49.055285
7  BYPU6 -0.400648  0.808337 -5.624670   72.943292
8  j9Oui  0.643105 -0.551583  2.372141  113.356160
9  OLuZU  2.173381  0.563698  9.441852  127.910945

In [3]:
data1.shape

(100000, 5)

In [4]:
data1.dtypes

id          object
f0         float64
f1         float64
f2         float64
product    float64
dtype: object

In [5]:
data1.isnull().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

We dont have any null value in Region 1 data

**Importing and studying Region 2 data**

In [6]:
data2 = pd.read_csv('/datasets/geo_data_1.csv')
data2.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [7]:
data2.shape

(100000, 5)

In [8]:
data2.dtypes

id          object
f0         float64
f1         float64
f2         float64
product    float64
dtype: object

In [9]:
data2.isnull().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

We dont have any null value in Region 2 data.

**Importing and studying Region 3 data**

In [10]:
data3 = pd.read_csv('/datasets/geo_data_2.csv')
data3.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

In [11]:
data3.dtypes

id          object
f0         float64
f1         float64
f2         float64
product    float64
dtype: object

In [12]:
data3.shape

(100000, 5)

In [13]:
data3.isnull().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

We dont have any null value in Region 3 data.

## Train and test the model for each region

### Train and test the model for region 1 :

In [14]:
target = data1['product']
features = data1.drop(['product','id'], axis=1)

In [15]:
#splitting data into 2 sets first: train = 60% and remaining=40%
features_train, features_valid, target_train, target_valid= train_test_split(features,target,test_size=0.25, random_state=12345)


In [16]:
model = LinearRegression().fit(features_train,target_train)
prediction = model.predict(features_valid)

In [17]:
prediction.mean()

92.59256778438038

In [18]:
print(model.coef_)
print(model.intercept_)

[  3.59280585 -14.09794419   6.59316617]
77.85671856415696


In [19]:
r2_score(prediction, target_valid)

-1.6290772227286001

In [20]:
MSE = mean_squared_error(prediction, target_valid)
#RMSE
MSE ** 0.5

37.5794217150813

### Train and test the model for region 2 :

In [21]:
target2 = data2['product']
features2 = data2.drop(['product','id'], axis=1)

In [22]:
#splitting data into 2 sets first: train = 60% and remaining=40%
features_train2, features_valid2, target_train2, target_valid2= train_test_split(features2,target2,test_size=0.25, random_state=12345)


In [23]:
model2 = LinearRegression().fit(features_train2,target_train2)
prediction2 = model2.predict(features_valid2)
prediction2.mean()

68.728546895446

In [24]:
r2_score(prediction2, target_valid2)

0.9996232023840883

In [25]:
MSE2 = mean_squared_error(prediction2, target_valid2)
#RMSE
MSE2 ** 0.5

0.893099286775616

### Train and test the model for region 3 :

In [26]:
target3 = data3['product']
features3 = data3.drop(['product','id'], axis=1)
#splitting data into 2 sets first: train = 60% and remaining=40%
features_train3, features_valid3, target_train3, target_valid3= train_test_split(features3,target3,test_size=0.25, random_state=12345)

model3 = LinearRegression().fit(features_train3,target_train3)
prediction3 = model3.predict(features_valid3)
prediction3.mean()


94.96504596800489

In [27]:

r2_score(prediction3, target_valid3)

-3.065349333298901

In [28]:
MSE3 = mean_squared_error(prediction3, target_valid3)
RMSE3 = MSE3 ** 0.5
RMSE3

40.02970873393434

**Conclusion**: Among the 3 models from 3 regions, we have the higest R-square score in region 2 data 0.999 as well as the value of RMSE is 0.89 which implies that the regression model fits our observations in region 2 data more accurately than other 2 regions.

## Prepare for profit calculation:

Calculate Break even point

In [34]:
#The budget for development of 200 oil wells
cost_well_build = 100000000

#revenue from one unit of product
revenue_per_unit = 4500
costof1cell= 100000000 / 200
break_even_point = costof1cell/4500
print(break_even_point)

111.11111111111111


## Profit and Risk Calculation

### Region 1 profit and risk calculation

In [30]:
target_valid1 = target_valid.reset_index(drop=True)
target_valid1 = pd.Series(target_valid1)
prediction1=pd.Series(prediction)

In [39]:

def revenue(target_valid1,prediction1,count):
    state = np.random.RandomState(12345)
    sorted_predictions = prediction1.sort_values(ascending=False) # sort prediction1
    target_sorted_by_preds = target_valid1[sorted_predictions.index][:count] # select target by sorted_predictions index
    target_sorted_by_preds_top200 =  target_sorted_by_preds[:200]# select first 200 elements from target_sorted_by_preds
    profit = ((target_sorted_by_preds_top200.sum()) * 4500) - 100000000
    return profit # sum of target_sorted_by_preds_top200 * revenue per unit - cost



state = np.random.RandomState(12345)
   
profit_list = []
for i in range(1000):
    target_subsample = target_valid1.sample(n=500,replace=True, random_state=state)
    probs_subsample = prediction1[target_subsample.index] 
    
    profit_list.append(revenue(target_subsample,probs_subsample,500))
                 
profit_list = pd.Series(profit_list)
lower = profit_list.quantile(0.01)
mean = profit_list.mean()
print("Average revenue:", mean)

neg_count = len(list(filter(lambda x: (x < 0), profit_list)))
risk_loss = (neg_count/1000)*100
print("The percentage of risk loss is", risk_loss, "%")

lower = profit_list.quantile(0.025)
upper = profit_list.quantile(0.975)


print("95% Confidence Interval:")
print("Lower quantile:", lower)
print("Upper quantile:", upper)

Average revenue: 4259385.269105923
The percentage of risk loss is 6.0 %
95% Confidence Interval:
Lower quantile: -1020900.9483793724
Upper quantile: 9479763.533583675


### Region 2 Profit and Risk Calculation

In [42]:
target_valid2 = target_valid2.reset_index(drop=True)
target_valid2 = pd.Series(target_valid2)
prediction2=pd.Series(prediction2)

def revenue2(target_valid2,prediction2,count):
    state = np.random.RandomState(12345)
    sorted_predictions = prediction2.sort_values(ascending=False) # sort prediction2
    target_sorted_by_preds = target_valid2[sorted_predictions.index][:count] # select target by sorted_predictions index
    target_sorted_by_preds_top200 =  target_sorted_by_preds[:200]# select first 200 elements from target_sorted_by_preds
    profit = (target_sorted_by_preds_top200.sum() * 4500) - 100000000
    return profit # sum of target_sorted_by_preds_top200 * revenue per unit - cost



state = np.random.RandomState(12345)
   
profit_list2 = []
for i in range(1000):
    target_subsample = target_valid2.sample(n=500,replace=True, random_state=state)
    probs_subsample = prediction2[target_subsample.index] 
    
    profit_list2.append(revenue2(target_subsample,probs_subsample,500))
                 
profit_list2 = pd.Series(profit_list2)
lower = profit_list2.quantile(0.01)
mean2 = profit_list2.mean()
print("Average revenue:", mean2)

neg_count2 = len(list(filter(lambda x: (x < 0), profit_list2)))
risk_loss2 = (neg_count2/1000)*100
print("The percentage of risk loss is", risk_loss2)

lower2 = profit_list2.quantile(0.025)
upper2 = profit_list2.quantile(0.975)


print("95% Confidence Interval:")
print("Lower quantile:", lower2)
print("Upper quantile:", upper2)

Average revenue: 5152227.734432899
The percentage of risk loss is 1.0
95% Confidence Interval:
Lower quantile: 688732.2537050088
Upper quantile: 9315475.912570495


### Region 3 Profit and Risk Calculation

In [41]:
target_valid3 = target_valid3.reset_index(drop=True)
target_valid3 = pd.Series(target_valid3)
prediction3=pd.Series(prediction3)

def revenue3(target_valid3,prediction3,count):
    state = np.random.RandomState(12345)
    sorted_predictions = prediction3.sort_values(ascending=False) # sort prediction3
    target_sorted_by_preds = target_valid3[sorted_predictions.index][:count] # select target by sorted_predictions index
    target_sorted_by_preds_top200 =  target_sorted_by_preds[:200]# select first 200 elements from target_sorted_by_preds
    profit = (target_sorted_by_preds_top200.sum() * 4500) - 100000000
    return profit # sum of target_sorted_by_preds_top200 * revenue per unit - cost



state = np.random.RandomState(12345)
   
profit_list3 = []
for i in range(1000):
    target_subsample = target_valid3.sample(n=500,replace=True, random_state=state)
    probs_subsample = prediction3[target_subsample.index] 
    
    profit_list3.append(revenue3(target_subsample,probs_subsample,500))
                 
profit_list3 = pd.Series(profit_list3)

mean3 = profit_list3.mean()
print("Average revenue:", mean3)

neg_count3 = len(list(filter(lambda x: (x < 0), profit_list3)))
risk_loss3 = (neg_count3/1000)*100
print("The percentage of risk loss is", risk_loss3)

lower3 = profit_list3.quantile(0.025)
upper3 = profit_list3.quantile(0.975)


print("95% Confidence Interval:")
print("Lower quantile:", lower3)
print("Upper quantile:", upper3)

Average revenue: 4350083.627827557
The percentage of risk loss is 6.4
95% Confidence Interval:
Lower quantile: -1288805.473297878
Upper quantile: 9697069.541802669


## Conclusion

- We can see we have the lowest risk of loss for Region 2 data which is 1%, while the risk of loss is ~6 % for Region 1 and Region 2.
- There is highest amount of average revenue generation for Region 2 data which is ~$ 5152227.
- Therefore, we can conclude that Region 2 is the most profitable option with lowest risk for oil mining than region 1 and region 3
